File: app/src/main/java/com/scop/org/minesweeper/utils/GridUtils.java
package com.scop.org.minesweeper.utils;

import com.scop.org.minesweeper.control.MainLogic;
import com.scop.org.minesweeper.control.CanvasWrapper;
import com.scop.org.minesweeper.control.GridDrawer;
import com.scop.org.minesweeper.elements.Grid;
import com.scop.org.minesweeper.elements.Tile;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.stream.Collectors;

public class GridUtils {



	public static List<Tile> getNeighbors(Grid grid, Tile tile){
		List<Tile> tiles = grid.getGrid();
		return getNeighborsIdx(grid, tile).stream()
				.map(tiles::get)
				.collect(Collectors.toList());
	}

	public static List<Integer> getNeighborsIdx(Grid grid, Tile tile){
		List<Integer> neighbors = new ArrayList<>();
		int w = grid.getW(),
			h = grid.getH(),
			idx = tile.getY()*w + tile.getX();

		// Test (w=7, h=3):
		// 0  1  2  3  4  5  6
		// 7  8  9 10 11 12 13
		//14 15 16 17 18 19 20

		boolean checkLeft = idx%w != 0;
		boolean checkTop = idx >= w;
		boolean checkRight = (idx+1)%w != 0;
		boolean checkBottom = idx < w*(h-1);

		if (checkLeft) {
			neighbors.add(idx-1); // L

			if (checkTop) {
				neighbors.add(idx-w); // T
				neighbors.add(idx-w-1); // TL
			}
			if (checkBottom) {
				neighbors.add(idx+w); // B
				neighbors.add(idx+w-1); // BL
			}
		}
		if (checkRight) {
			neighbors.add(idx+1); // R

			if (checkTop) {
				if (!checkLeft) neighbors.add(idx-w); // T
				neighbors.add(idx-w+1); // TR
			}
			if (checkBottom) {
				if (!checkLeft) neighbors.add(idx+w); // B
				neighbors.add(idx+w+1); // BR
			}
		}
		else if (!checkLeft) {
			if (checkTop)    neighbors.add(idx-w); // T
			if (checkBottom) neighbors.add(idx+w); // B
		}
		return neighbors;
	}

	public static Tile getTileByScreenCoords(Grid grid, float x, float y){
		int tileSize = GridDrawer.getTileSize();
		float scale = CanvasWrapper.getScale();
		float posX = CanvasWrapper.getPosX();
		float posY = CanvasWrapper.getPosY();

		return getTileByCoords(grid,
				Math.round((-posX + x) / scale) / tileSize,
				Math.round((-posY + y) / scale) / tileSize);
	}

	public static Tile getTileByCoords(Grid grid, int x, int y){
		int w = grid.getW(),
			idx = y*w + x;

		if (idx >= 0 && idx < grid.getGrid().size())
			return grid.getGrid().get(idx);

		return null;
	}

	public static Tile findSafeOpenTile(Grid grid) {
		Random rnd = new Random();
		List<Tile> tiles = grid.getGrid();
		Tile t;
		do {
			t = tiles.get(rnd.nextInt(tiles.size()));

		} while (t.hasBomb() || t.getBombsNear() > 0 || t.getStatus() != Tile.Status.COVERED);

		return t;
	}

	public static int findSafeOpenTileIdx(Grid grid) {
		Random rnd = new Random();
		List<Tile> tiles = grid.getGrid();
		Tile t;
		int idx;
		do {
			idx = rnd.nextInt(tiles.size());
			t = tiles.get(idx);

		} while (t.hasBomb() || t.getBombsNear() > 0 || t.getStatus() != Tile.Status.COVERED);

		return idx;
	}

	public static int getTileNumber(Tile tile){
		switch (tile.getStatus()){
			case A1: return 1;
			case A2: return 2;
			case A3: return 3;
			case A4: return 4;
			case A5: return 5;
			case A6: return 6;
			case A7: return 7;
			case A8: return 8;
		}
		return -1;
	}
	public static Tile.Status getTileStatus(int value){
		switch (value){
			case 1: return Tile.Status.A1;
			case 2: return Tile.Status.A2;
			case 3: return Tile.Status.A3;
			case 4: return Tile.Status.A4;
			case 5: return Tile.Status.A5;
			case 6: return Tile.Status.A6;
			case 7: return Tile.Status.A7;
			case 8: return Tile.Status.A8;
		}
		return Tile.Status.A0;
	}


	public static JSONObject getJsonStatus(Grid grid, int seconds) {
		JSONObject obj = new JSONObject();
		try {
			obj.put("w", grid.getW());
			obj.put("h", grid.getH());
			obj.put("g", grid.getGeneratorClass());
			obj.put("x", CanvasWrapper.getPosX());
			obj.put("y", CanvasWrapper.getPosY());
			obj.put("s", CanvasWrapper.getScale());
			obj.put("t", seconds);
			obj.put("ts",
					grid.getGrid().stream()
						.map(t->{
							String c;
							switch(t.getStatus()) {
								case BOMB: c = "B"; break;
								case FLAG: c = t.hasBomb()? "F":"f"; break;
								case COVERED: c = t.hasBomb()? "C":"c"; break;
								case A1: c = "1"; break;
								case A2: c = "2"; break;
								case A3: c = "3"; break;
								case A4: c = "4"; break;
								case A5: c = "5"; break;
								case A6: c = "6"; break;
								case A7: c = "7"; break;
								case A8: c = "8"; break;
								default:
								case A0: c = " "; break;
							}
							return c;
						})
						.collect(Collectors.joining())
			);
		} catch (JSONException e) {
			e.printStackTrace();
		}
		return obj;
	}

	public static MainLogic calculateLogicFromBareGrid(Grid grid) {
		MainLogic logic = new MainLogic(grid);

		for (Tile t : grid.getGrid()) {
			if (t.hasBomb()) {
				GridUtils.getNeighbors(grid, t).forEach(Tile::hasBombNear);
			}

			if (!t.isCovered()) {
				logic.addRevealedTiles();
			}
			if (t.getStatus() == Tile.Status.FLAG){
				GridUtils.getNeighbors(grid, t).forEach(Tile::addFlaggedNear);

				logic.addFlaggedBombs();
				if (t.hasBomb()) {
					logic.addCorrectFlaggedBombs();
				}
			}
		}

		return logic;
	}
}

File: app/src/main/java/org/oar/minesweeper/utils/GridUtils.kt
package org.oar.minesweeper.utils

import android.content.Context
import org.json.JSONException
import org.json.JSONObject
import org.oar.minesweeper.grid.GridPosition
import org.oar.minesweeper.grid.GridDrawer.tileSize
import org.oar.minesweeper.grid.GameLogic
import org.oar.minesweeper.models.Grid
import org.oar.minesweeper.models.*
import org.oar.minesweeper.models.TileStatus.*
import org.oar.minesweeper.utils.PreferencesUtils.loadBoolean
import java.util.*
import kotlin.math.roundToInt
import kotlin.reflect.full.createInstance

object GridUtils {

    fun Grid.generate(callback: (GridGenerationDetails) -> Unit) {
        tiles.clear()
        val gen = gridSettings.generatorClass.createInstance()
        gen.generateNewGrid(this, callback)
    }

    fun Grid.getNeighbors(tile: Tile): List<Tile> {
        val tiles: List<Tile> = tiles
        return getNeighborsIdx(tile)
            .map { idx -> tiles[idx] }
    }

    fun Grid.getNeighborsIdx(tile: Tile): MutableList<Int> {
        return gridConfig.getNeighborsIdx(tile)
    }

    fun GridConfiguration.getNeighborsIdx(tile: Tile): MutableList<Int> {
        val neighbors: MutableList<Int> = ArrayList()
        val idx = tile.y * width + tile.x

        // Test (w=7, h=3):
        // 0  1  2  3  4  5  6
        // 7  8  9 10 11 12 13
        //14 15 16 17 18 19 20

        val checkLeft = idx % width != 0
        val checkTop = idx >= width
        val checkRight = ((idx + 1) % width) != 0
        val checkBottom = idx < width * (height - 1)
        if (checkLeft) {
            neighbors.add(idx - 1) // L
            if (checkTop) {
                neighbors.add(idx - width) // T
                neighbors.add(idx - width - 1) // TL
            }
            if (checkBottom) {
                neighbors.add(idx + width) // B
                neighbors.add(idx + width - 1) // BL
            }
        }
        if (checkRight) {
            neighbors.add(idx + 1) // R
            if (checkTop) {
                if (!checkLeft) neighbors.add(idx - width) // T
                neighbors.add(idx - width + 1) // TR

            }
            if (checkBottom) {
                if (!checkLeft) neighbors.add(idx + width) // B
                neighbors.add(idx + width + 1) // BR
            }
        } else if (!checkLeft) {
            if (checkTop) { // T
                neighbors.add(idx - width)
            }
            if (checkBottom) { // B
                neighbors.add(idx + width)
            }
        }
        return neighbors
    }


    fun Grid.getTileByScreenCoords(x: Float, y: Float, gridPosition: GridPosition): Tile? {
        val scale = gridPosition.scale
        val relativeX = (x - gridPosition.posX) / scale
        val relativeY = (y - gridPosition.posY) / scale

        if (relativeX < 0 || relativeY < 0) return null

        return getTileByCoords(
            relativeX.roundToInt() / tileSize,
            relativeY.roundToInt() / tileSize)
    }

    fun Grid.getTileByCoords(x: Int, y: Int): Tile? {
        if (x >= width) return null

        val idx = y * width + x
        return if (idx >= 0 && idx < tiles.size) tiles[idx] else null
    }

    fun Grid.findSafeOpenTile(): Tile {
        return tiles[findSafeOpenTileIdx()]
    }

    fun Grid.findSafeOpenTileIdx(): Int {
        val rnd = Random()
        val tiles = tiles
        var idx: Int
        do {
            idx = rnd.nextInt(tiles.size)
            val tile = tiles[idx]
        } while (tile.hasBomb || tile.bombsNear > 0 || tile.status !== COVERED)
        return idx
    }

    fun Grid.generateLogic(): GameLogic {
        val logic = GameLogic(this)

        tiles.forEach { tile ->
            if (tile.hasBomb) {
                getNeighbors(tile).forEach { it.bombsNear++ }
            }
            if (!tile.isCovered) {
                logic.addRevealedTiles()
            }
            if (tile.status === FLAG) {
                getNeighbors(tile).forEach { it.flaggedNear++ }
                logic.addFlaggedBombs()
                if (tile.hasBomb) {
                    logic.addCorrectFlaggedBombs()
                }
            }
        }

        return logic
    }

    fun Grid.toJson(deciSeconds: Int, gridPosition: GridPosition): JSONObject {
        val jsonObj = JSONObject()

        try {
            jsonObj.put("w", width)
            jsonObj.put("h", height)
            jsonObj.put("gs", gridSettings.solvable)
            jsonObj.put("x", gridPosition.posX)
            jsonObj.put("y", gridPosition.posY)
            jsonObj.put("s", gridPosition.scale)
            jsonObj.put("t", deciSeconds)
            jsonObj.put("ts",
                tiles
                    .joinToString("") { tile ->
                        when (tile.status) {
                            BOMB -> "B"
                            FLAG -> if (tile.hasBomb) "F" else "f"
                            COVERED -> if (tile.hasBomb) "C" else "c"
                            A1, A2, A3, A4, A5, A6, A7, A8 -> tile.status.tileNumber.toString()
                            else -> " "
                        }
                    }
            )
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return jsonObj
    }

    fun Context.gridFromJson(
        jsonObj: JSONObject,
        gridSettings: GridSettings? = null
    ): Grid {
        val height = jsonObj.getInt("h")
        val width = jsonObj.getInt("w")
        val tilesList: MutableList<Tile> = mutableListOf()
        var bombs = 0
        val stringGrid = jsonObj.getString("ts")

        for (i in stringGrid.indices) {
            var t: Tile? = null
            when (stringGrid[i]) {
                'B' -> t = Tile(i % width, i / width, TileStatus.BOMB)
                ' ' -> t = Tile(i % width, i / width, TileStatus.A0)
                '1' -> t = Tile(i % width, i / width, TileStatus.A1)
                '2' -> t = Tile(i % width, i / width, TileStatus.A2)
                '3' -> t = Tile(i % width, i / width, TileStatus.A3)
                '4' -> t = Tile(i % width, i / width, TileStatus.A4)
                '5' -> t = Tile(i % width, i / width, TileStatus.A5)
                '6' -> t = Tile(i % width, i / width, TileStatus.A6)
                '7' -> t = Tile(i % width, i / width, TileStatus.A7)
                '8' -> t = Tile(i % width, i / width, TileStatus.A8)
                'f' -> t = Tile(i % width, i / width, TileStatus.FLAG)
                'c' -> t = Tile(i % width, i / width, TileStatus.COVERED)
                'F' -> {
                    t = Tile(i % width, i / width, TileStatus.FLAG)
                        .apply { hasBomb = true }
                    bombs++
                }
                'C' -> {
                    t = Tile(i % width, i / width, TileStatus.COVERED)
                        .apply { hasBomb = true }
                    bombs++
                }
            }
            if (t != null) tilesList.add(t)
        }

        val isSolvable = if (jsonObj.has("gs"))
            jsonObj.getBoolean("gs")
        else
            false

        val config = GridConfiguration(width, height, bombs)
        val settings = gridSettings ?: GridSettings(
            loadBoolean("lastRevealFirst", true),
            isSolvable,
            loadBoolean("lastVisualHelp", false),
        )

        return Grid(config, settings).apply {
            tiles = tilesList
        }
    }
}